libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <vector>
+#include <utility>
+
+namespace fc {
+
+/**
+ * @class ordered_diff
+ * @brief Provides ability to generate and apply diff of containers of type T
+ *
+ * Example use:
+ *    std::vector<char> source = { 'a', 'b', 'f', 'c', 'd' };
+ *    std::vector<char> target = { 'b', 'f', 'c', 'd', 'e', 'h' };
+ *    ordered_diff<char>::diff_result diff = ordered_diff<char>::diff(source, target);
+ *    ordered_diff<char>::apply_diff(source, std::move(diff));
+ *    assert(source == target);
+ *
+ * @param T type stored in Container, must provide ==
+ * @param SizeType numeric type used for index into diff_result, for non-unique Containers a larger type may be required
+ * @param Container container type for ordered diff and for diff_result
+ */
+template <typename T, typename SizeType = size_t, template<typename Y, typename...> typename Container = std::vector>
+requires std::equality_comparable<T> && std::random_access_iterator<typename Container<T>::iterator>
+class ordered_diff {
+public:
+   struct diff_result {
+      Container<SizeType>                remove_indexes;
+      Container<std::pair<SizeType, T>>  insert_indexes;
+   };
```

Comment:
Why not have `apply_diff` here as a member function

`Container<T> apply(Container<T>&& source) &&;`

Reply:
It might be a bit too clever, but the current implementation works for both a `const diff_result` and a rvalue `diff_result`. Not sure if it is possible to share implementation between a const member function and a `&&` member function.

Reply:
Ah, yes, makes sense. We'll be able to support this in a member finction with c++23 [deducing_this](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0847r6.html#deduplicating-code).